tf2onnx/rewriter/gemm_rewriter.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT license.
+
+"""
+tf2onnx.rewrite - rewrite tensorflow subgraph to onnx gemm op
+"""
+import logging
+from onnx import onnx_pb
+from tf2onnx.graph_matcher import OpTypePattern, GraphMatcher
+
+# pylint: disable=missing-docstring
+
+def rewrite_gemm(g, ops):
+    if g.opset <= 6:
+        return ops
+
+    # pattern0: alpha*A*B + beta*C
+    pattern0 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('Mul', name='mul1', inputs=[
+                OpTypePattern('Const', name='alpha'),
+                OpTypePattern('MatMul', name='matmul')
+            ]),
+            OpTypePattern('Mul', name='mul2', inputs=[
+                OpTypePattern('Const', name='beta'),
+                OpTypePattern('*', name='C')
+            ])
+        ])
+
+    # pattern1: alpha*A*B + C
+    pattern1 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('Mul', name='mul1', inputs=[
+                OpTypePattern('MatMul', name='matmul'),
+                OpTypePattern('Const', name='alpha')
+            ]),
+            OpTypePattern('*', name='C'),
+        ])
+
+    # pattern2: A*B + beta*C
+    pattern2 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('MatMul', name='matmul'),
+            OpTypePattern('Mul', name='mul2', inputs=[
+                OpTypePattern('Const', name='beta'),
+                OpTypePattern('*', name='C')
+            ])
+        ])
+
+    # pattern3: A*B + C
+    pattern3 = \
+        OpTypePattern('Add', name='add', inputs=[
+            OpTypePattern('MatMul', name='matmul'),
+            OpTypePattern('*', name='C'),
+        ])
+
+    pattern_list = [pattern0, pattern1, pattern2, pattern3]
+
+    for pattern_id, pattern in enumerate(pattern_list):
+        matcher = GraphMatcher(pattern, allow_reorder=True)
+        match_results = list(matcher.match_ops(ops))
+        if match_results:
+            for match in match_results:
+                add_node = match.get_op('add')
+                matmul_node = match.get_op("matmul")
+                input_c_node = match.get_op("C")
+
+                if g.get_dtype(matmul_node.input[0]) != onnx_pb.TensorProto.FLOAT:
+                    logging.warning(u"For now, onnxruntime only support float type for Gemm rewriter")
+                    return ops
```

Comment:
what if other match result can be rewrited to gemm??

Reply:
Very appriciate. I have use continue instead of return ops.

Reply:
please be careful...

Reply:
So sorry for the error.